✨ Feat : 로그인 API 추가

seunGit opened the pull request:
<!-- 
체크하려면 괄호 안에 "x"를 입력하세요. 
각 규칙은 Convention 문서에 있습니다.
PR 제목에 쓰는 prefix는 다음과 같습니다.

✨ Feat : 새로운 기능
🔨️ Refactor : 코드 리팩토링
🐎 Perf : 성능을 향상
🐛 Fix : 버그를 고칠 때
🧪 Test : 테스트 코드
🚜 Rename : 파일 이름 변경 혹은 구조를 변경
🚀 Release : 배포 / 개발 작업과 관련된 모든 것
🔥 Remove : 코드 또는 파일 제거
📚 Docs : 문서
📝 Chore : 사소한 코드 또는 언어를 변경 기타 변경사항 (빌드 스크립트 수정, 패키지 매니징 설정 등)

-->

## PR 타입
<!-- 어떤 유형의 PR인지 체크해주세요. -->

<!-- 체크하려면 괄호 안에 "x"를 입력하세요. -->
- [x] Feature
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Performance Improvement
- [ ] Bugfix
- [x] Test Code
- [ ] Code style update (formatting, local variables)
- [ ] Other... Please describe:



## 개요 
- 로그인 기능 추가
- 구현된 코드를 테스트하는 테스트 코드 추가

## 작업 및 변경 사항 
- 로그인 기능 구현
- Repository계층 단위 테스트 코드 추가
- Service계층 단위 테스트 코드 추가
- Controller계층 통합 테스트 코드 추가

### 테스트 결과 
- 테스트 코드 실행 특이사항 없음

close #9

seunGit committed:
[test] 로그인기능 테스트 코드 추가.
Repository, Service 계층 단위테스트 코드 추가.
Controller 계층 은 통합테스트 진행.
로그인 기능 테스트 완료.

Resolves: #9

seunGit committed:
[feat] 로그인 기능 추가.
1. 로그인 기능 추가

Resolves: #9

m1naworld commented on src/test/java/com/hanbang/e/member/repository/MemberRepositoryTest.java:
```diff
@@ -16,6 +18,21 @@ public class MemberRepositoryTest {
+        Member member2 = new Member("이메일1","비밀번호1","주소1");
+        memberRepository.save(member2);
+    }
+
+    @AfterEach
+    public void clear() {
+        memberRepository.deleteAll();
+    }
+
```
제가 알기로는 `@DataJpaTest` 어노테이션 안에 `@Transactional`에 의해 하나의 테스트 메서드가 종료될 때마다 테스트 메서드 내부에서 생성했던 데이터를 rollback 한다고 알고 있습니다. ❗️ 혹시 이를 적용하고 안하고의 차이를 알 수 있을까요?! 혹시나 안할 경우 어떠한 에러가 나는지도 궁금합니다👀

seunGit replied:
`AfterEach` 사용유무와 상관없이 실행이 잘 되는걸 확인했습니다. <br> 말씀하신대로 `@DataJpaTest`  내의  `@Transactional`이 작동하여 필요없는 코드로 생각됩니다. 추후 학습 후 말씀드릴 수 있도록 하겠습니다.

m1naworld reviewed (commented):
고생하셨습니다 👏🏻

m1naworld reviewed (commented):
현재 jwt를 검증하는 로직은 없는 것 같은데 구현 중이신걸까요?!

hyunmuam commented on src/main/java/com/hanbang/e/member/service/MemberService.java:
```diff
@@ -10,20 +12,37 @@
         memberRepository.save(member);
         return MemberResp.from(member);
     }
+
+    @Transactional(readOnly = true)
+    public MemberResp login(MemberLoginReq memberLoginReq, HttpServletResponse httpServletResponse) {
+        Member member = memberRepository.findByEmail(memberLoginReq.getEmail()).orElseThrow(
+                () -> new IllegalArgumentException("등록된 사용자가 없습니다.")
+        );
+        if (!member.getPassword().equals(memberLoginReq.getPassword())) {
+            throw new IllegalArgumentException("비밀번호가 일치하지 않습니다.");
+        }
+        httpServletResponse.addHeader(JwtUtil.AUTHORIZATION_HEADER, jwtUtil.createToken(member.getEmail()));
+        return MemberResp.from(member);
+    }
```
저희 API명세서의 반환타입이랑 다르게 구현이 되어있네요. API명세서 확인해주시고 API명세서나 본 코드를 수정하던가 해야될 듯 합니다.

seunGit replied:
@tjvm0877 API 명세서 반환타입은 컨트롤러 계층에 구현했습니다. Service 계층의 반환타입은 ServiceTest코드 작성을 위해 구현되었습니다.

hyunmuam commented on src/test/java/com/hanbang/e/member/controller/MemberControllerTest.java:
```diff
@@ -61,4 +81,50 @@ public void signupTest() throws JsonProcessingException {
+        MemberLoginReq memberLoginReq1 = new MemberLoginReq("이메일" ,"비밀번호");
+        String body1 = om.writeValueAsString(memberLoginReq1);
+
+        MemberLoginReq memberLoginReq2 = new MemberLoginReq("이메일1" ,"비밀번호");
+        String body2 = om.writeValueAsString(memberLoginReq2);
+
+        MemberLoginReq memberLoginReq3 = new MemberLoginReq("이메일" ,"비밀번호1");
+        String body3 = om.writeValueAsString(memberLoginReq3);
+
+        /* when - 테스트 실행 */
+        HttpEntity<String> request1 = new HttpEntity<>(body1, headers);
+        ResponseEntity<String> response1 = rt.exchange("/api/member/login", HttpMethod.POST, request1, String.class);
+
+        HttpEntity<String> request2 = new HttpEntity<>(body2, headers);
+        ResponseEntity<String> response2 = rt.exchange("/api/member/login", HttpMethod.POST, request2, String.class);
+
+        HttpEntity<String> request3 = new HttpEntity<>(body3, headers);
+        ResponseEntity<String> response3 = rt.exchange("/api/member/login", HttpMethod.POST, request3, String.class);
+
+        /* then - 검증 */
+        DocumentContext dc1 = JsonPath.parse(response1.getBody());
+        String result1 = dc1.read("$.result");
+        String msg1 = dc1.read("$.msg");
+
+        DocumentContext dc2 = JsonPath.parse(response2.getBody());
+        String result2 = dc2.read("$.result");
+        String msg2 = dc2.read("$.msg");
+
+        DocumentContext dc3 = JsonPath.parse(response3.getBody());
+        String result3 = dc3.read("$.result");
+        String msg3 = dc3.read("$.msg");
+
+        assertThat(result1).isEqualTo("success");
+        assertThat(msg1).isEqualTo("로그인 완료");
+
+        assertThat(result2).isEqualTo("fail");
+        assertThat(msg2).isEqualTo("등록된 사용자가 없습니다.");
+
+        assertThat(result3).isEqualTo("fail");
+        assertThat(msg3).isEqualTo("비밀번호가 일치하지 않습니다.");
```
하나의 테스트안에 이렇게놔도 괜찮아보이네요. 저는 전부다 분리하긴 했는데 고민해볼 여지가 있는 것 같습니다. 다른분들의 의견이 궁금합니다.

m1naworld replied:
저도 모두 분리했는데 🤣 그래서 중복되는 코드가 많습니다. 생각해보면 하나의 기능을 검증하는 테스트인데 결과에 따라 여러 테스트로 나눈 다는 것이.. 🤔 위의 코드가 기능 별 예외처리 및 직관적으로 보여 더 좋아보입니다. 테스트 코드 리팩토링시 우선 하나로 통합하는 방향으로 가보는건 어떨까요?!

hyunmuam replied:
@m1naworld 근데 또 생각해보면 하나의 기능에 대한 테스트코드가 많아질수록 점점 단위테스트의 의미가 퇴색될 수 있다는 점이 있을 것 같네요.

hyunmuam reviewed (approved):
저희가 짠 API명세서의 내용과 로그인 기능이 실제로 반환하는 내용이 차이가 있는 것 같습니다. 확인 부탁드리고 명세서 혹은 코드를 수정해서 맞춰주시면 될 듯 합니다.
제 개인적인 의견이지만 로그인할 때 주소와 이메일은 반환할 필요가 없다고 생각합니다.

seunGit commented on src/main/java/com/hanbang/e/common/jwt/JwtUtil.java:
```diff
@@ -0,0 +1,86 @@
+                        .compact();
+    }
+
+    public boolean validateToken(String token) {
+        try {
+            Jwts.parserBuilder().setSigningKey(key).build().parseClaimsJws(token);
+            return true;
+        } catch (SecurityException | MalformedJwtException e) {
+            log.info("Invalid JWT signature, 유효하지 않는 JWT 서명 입니다.");
+        } catch (ExpiredJwtException e) {
+            log.info("Expired JWT token, 만료된 JWT token 입니다.");
+        } catch (UnsupportedJwtException e) {
+            log.info("Unsupported JWT token, 지원되지 않는 JWT 토큰 입니다.");
+        } catch (IllegalArgumentException e) {
+            log.info("JWT claims is empty, 잘못된 JWT 토큰 입니다.");
+        }
+        return false;
+    }
```
@m1naworld 토큰 검증하는 단계 입니다.

hyunmuam reviewed (changes requested):
잘못 승인 했습니다. 코멘트 확인해주세요 ㅠ

seunGit committed:
[test] 로그인기능 테스트 코드 수정.
Repository, Service, Controller 계층 단위테스트 코드 수정.
Controller 계층은 통합테스트 진행.
jwtUtilTest 추가
로그인 기능 테스트 완료.

Resolves: #9

seunGit committed:
[refactor] 로그인 기능 코드 수정
- toEntity() 메서드 추가
- MemberService 반환타입 수정

Resolves: #9

m1naworld commented on src/test/java/com/hanbang/e/member/util/JwtUtilTest.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    @Test
+    public void createToken() {
+        /* given - 데이터 준비 */
+        MemberLoginReq memberLoginReq = new MemberLoginReq(
+                "이메일", "비밀번호");
+        MemberCreateReq memberCreateReq = new MemberCreateReq(
+                "이메일1", "비밀번호1","주소1");
+
+        /* token 생성 */
+        String token1 = this.jwtUtil.createToken(memberLoginReq.getEmail());
+        String token2 = this.jwtUtil.createToken(memberCreateReq.getEmail());
+
+        /* then - 검증 */
+        assertThat(token1).isNotNull();
+        assertThat(token1).isNotEqualTo(token2).isNotNull();
+    }
```
좋습니다!

m1naworld commented on src/test/java/com/hanbang/e/member/service/MemberServiceTest.java:
```diff
@@ -46,15 +62,60 @@ public void signupTest() {
     public void loginTest() {
         /* given - 데이터 준비 */
-        MemberCreateReq memberCreateReq = new MemberCreateReq(
+        HttpServletResponse httpServletResponse = mock(HttpServletResponse.class);
+        Member member = new Member(
                 "이메일", "비밀번호", "주소");
         MemberLoginReq memberLoginReq = new MemberLoginReq(
                 "이메일", "비밀번호");
+
+        /* stub - 가짜 객체 행동 정의 */
+        when(memberRepository.findByEmail(any())).thenReturn(Optional.of((memberLoginReq.toEntity())));
```
stub 부분에서 memberLoginReq.toEntity가 아닌 member가 와야할 것 같습니다. ! 그 이유에 대해 한번 생각해보시고 다음 리팩토링 때 참고 부탁드립니다 🙇‍♀️ 또한, 현재 httpServletResponse는 사용하지 않는 코드로 보입니다. 이런 경우에는 지워주시면 감사하겠습니다! 🫡

m1naworld reviewed (approved):
고생하셨습니다 👏🏻

seunGit committed:
Merge branch 'develop' into feature/#9
